Marlin/src/module/tool_change.cpp

```diff
@@ -508,7 +513,6 @@ void tool_change(const uint8_t tmp_extruder, const float fr_mm_s/*=0.0*/, bool n
             prepare_move_to_destination();
           }
```

Comment:
This code is left in place for the case where `no_move` is true or `IsRunning()` is false, in which case the Z move represented here wouldn't have occurred yet. As noted above, if the move _was_ done above as part of "move back," then this is prevented from happening.

Reply:
why not
        #if ENABLED(SWITCHING_NOZZLE)
          destination[Z_AXIS] += z_diff;  // Include the Z restore with the "move back"
        #endif
        if (!no_move && IsRunning()) {
          #if ENABLED(DEBUG_LEVELING_FEATURE)
            if (DEBUGGING(LEVELING)) DEBUG_POS("Move back", destination);
          #endif
          // Move back to the original (or tweaked) position
          do_blocking_move_to(destination[X_AXIS], destination[Y_AXIS], destination[Z_AXIS]);
        }
       
        #if ENABLED(SWITCHING_NOZZLE)
          else {
            // Move back down, if needed. (Including when the new tool is higher.)
            feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS];
	    do_blocking_move_to(destination[X_AXIS], destination[Y_AXIS], destination[Z_AXIS]);
          }
        #endif

Reply:
but note... zx& Y should not move if no home...this is still missing (I'm always talkin of switching_nozzle option)

Reply:
Ah, very true. In that case, before the Z move for `SWITCHING_NOZZLE`, it needs to `set_destination_from_current`, so the planner and cartesian position are in sync prior to the move.

Reply:
Sorry this is so hard to get straight. I haven't got any multi-nozzle setups, and I have to work it all out in my head. So, with that consideration in mind…
```cpp
// Move to the "old position" (move the extruder into place)
if (!no_move && IsRunning()) {
  #if ENABLED(SWITCHING_NOZZLE)
    if (z_raise != z_diff)
      destination[Z_AXIS] += z_diff;  // Include the Z restore with the "move back"
  #endif
  #if ENABLED(DEBUG_LEVELING_FEATURE)
    if (DEBUGGING(LEVELING)) DEBUG_POS("Move back", destination);
  #endif
  // Move back to the original (or tweaked) position
  do_blocking_move_to(destination[X_AXIS], destination[Y_AXIS], destination[Z_AXIS]);
}
#if ENABLED(SWITCHING_NOZZLE)
  // Move back down, if needed. (Including when the new tool is higher.)
  else if (z_raise != z_diff) {
    set_destination_from_current(); // Prevent any XY move
    destination[Z_AXIS] += z_diff;
    feedrate_mm_s = planner.max_feedrate_mm_s[Z_AXIS];
    prepare_move_to_destination();
  }
#endif
```